Errors in log caused by switching to trace mode

I was trying to help a user on the [forum](https://www.universalmediaserver.com/forum/viewtopic.php?f=9&t=15320) and found database initialization and unable to bind errors in the logs. I have noticed the same thing from other users the past few versions.

I also noticed that most of the recent logs are not in trace mode at the start. Then UMS is switched to trace mode and the logs continue showing UMS starting up but that is when these errors show up.

I think when UMS is put into trace mode, some of the old processes are still shutting down when the new ones are started so things like the database cannot be initialized.

The log file is not refreshed like it used to be and is just appended to the end of the old one instead of the old one renamed with prev and a new one started.

Here are a few other logs that show this from the first page on the forum and I'm sure there are more going back further.
https://www.universalmediaserver.com/forum/viewtopic.php?p=45984#p45984
https://www.universalmediaserver.com/forum/viewtopic.php?p=46044#p46044
https://www.universalmediaserver.com/forum/viewtopic.php?p=45797#p45797

[this](https://www.universalmediaserver.com/forum/viewtopic.php?p=46159#p46159) one shows UMS starting 14 times

and the errors that are similar are
```
ERROR 2022-09-14 18:11:10.578 [AWT-EventQueue-0] Can´t find Java executable "C:\Program Files (x86)\Universal Media Server\jre17\bin\javaw", falling back to pathless execution using "javaw"
ERROR 2022-09-14 18:11:17.165 [main] com.zaxxer.hikari.pool.HikariPool HikariPool-1 - Exception during pool initialization.
org.h2.jdbc.JdbcSQLNonTransientConnectionException: Database may be already in use: "C:/ProgramData/UMS/database/medias.mv.db". Possible solutions: close all other connection(s); use the server mode [90020-214]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:678)
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:477)
	at org.h2.message.DbException.get(DbException.java:212)
	at org.h2.mvstore.db.Store.convertMVStoreException(Store.java:168)
	at org.h2.mvstore.db.Store.<init>(Store.java:145)
	at org.h2.engine.Database.<init>(Database.java:324)
	at org.h2.engine.Engine.openSession(Engine.java:92)
	at org.h2.engine.Engine.openSession(Engine.java:222)
	at org.h2.engine.Engine.createSession(Engine.java:201)
	at org.h2.engine.SessionRemote.connectEmbeddedOrServer(SessionRemote.java:338)
	at org.h2.jdbc.JdbcConnection.<init>(JdbcConnection.java:122)
	at org.h2.Driver.connect(Driver.java:59)
	at com.zaxxer.hikari.util.DriverDataSource.getConnection(DriverDataSource.java:138)
	at com.zaxxer.hikari.pool.PoolBase.newConnection(PoolBase.java:359)
	at com.zaxxer.hikari.pool.PoolBase.newPoolEntry(PoolBase.java:201)
	at com.zaxxer.hikari.pool.HikariPool.createPoolEntry(HikariPool.java:470)
	at com.zaxxer.hikari.pool.HikariPool.checkFailFast(HikariPool.java:561)
	at com.zaxxer.hikari.pool.HikariPool.<init>(HikariPool.java:100)
	at com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:112)
	at net.pms.database.Database.getConnection(Database.java:74)
	at net.pms.database.Database.open(Database.java:111)
	at net.pms.database.Database.init(Database.java:99)
	at net.pms.database.MediaDatabase.init(MediaDatabase.java:189)
	at net.pms.PMS.init(PMS.java:418)
	at net.pms.PMS.createInstance(PMS.java:955)
	at net.pms.PMS.main(PMS.java:1120)
Caused by: org.h2.mvstore.MVStoreException: The file is locked: C:/ProgramData/UMS/database/medias.mv.db [2.1.214/7]
	at org.h2.mvstore.DataUtils.newMVStoreException(DataUtils.java:1004)
	at org.h2.mvstore.FileStore.open(FileStore.java:178)
	at org.h2.mvstore.FileStore.open(FileStore.java:128)
	at org.h2.mvstore.MVStore.<init>(MVStore.java:452)
	at org.h2.mvstore.MVStore$Builder.open(MVStore.java:4082)
	at org.h2.mvstore.db.Store.<init>(Store.java:136)
	... 21 common frames omitted
ERROR 2022-09-14 18:11:28.203 [main] com.zaxxer.hikari.pool.HikariPool HikariPool-1 - Exception during pool initialization.
org.h2.jdbc.JdbcSQLNonTransientConnectionException: Database may be already in use: "C:/ProgramData/UMS/database/medias.mv.db". Possible solutions: close all other connection(s); use the server mode [90020-214]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:678)
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:477)
	at org.h2.message.DbException.get(DbException.java:212)
	at org.h2.mvstore.db.Store.convertMVStoreException(Store.java:168)
	at org.h2.mvstore.db.Store.<init>(Store.java:145)
	at org.h2.engine.Database.<init>(Database.java:324)
	at org.h2.engine.Engine.openSession(Engine.java:92)
	at org.h2.engine.Engine.openSession(Engine.java:222)
	at org.h2.engine.Engine.createSession(Engine.java:201)
	at org.h2.engine.SessionRemote.connectEmbeddedOrServer(SessionRemote.java:338)
	at org.h2.jdbc.JdbcConnection.<init>(JdbcConnection.java:122)
	at org.h2.Driver.connect(Driver.java:59)
	at com.zaxxer.hikari.util.DriverDataSource.getConnection(DriverDataSource.java:138)
	at com.zaxxer.hikari.pool.PoolBase.newConnection(PoolBase.java:359)
	at com.zaxxer.hikari.pool.PoolBase.newPoolEntry(PoolBase.java:201)
	at com.zaxxer.hikari.pool.HikariPool.createPoolEntry(HikariPool.java:470)
	at com.zaxxer.hikari.pool.HikariPool.checkFailFast(HikariPool.java:561)
	at com.zaxxer.hikari.pool.HikariPool.<init>(HikariPool.java:100)
	at com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:112)
	at net.pms.database.Database.getConnection(Database.java:74)
	at net.pms.database.Database.open(Database.java:124)
	at net.pms.database.Database.init(Database.java:99)
	at net.pms.database.MediaDatabase.init(MediaDatabase.java:189)
	at net.pms.PMS.init(PMS.java:418)
	at net.pms.PMS.createInstance(PMS.java:955)
	at net.pms.PMS.main(PMS.java:1120)
Caused by: org.h2.mvstore.MVStoreException: The file is locked: C:/ProgramData/UMS/database/medias.mv.db [2.1.214/7]
	at org.h2.mvstore.DataUtils.newMVStoreException(DataUtils.java:1004)
	at org.h2.mvstore.FileStore.open(FileStore.java:178)
	at org.h2.mvstore.FileStore.open(FileStore.java:128)
	at org.h2.mvstore.MVStore.<init>(MVStore.java:452)
	at org.h2.mvstore.MVStore$Builder.open(MVStore.java:4082)
	at org.h2.mvstore.db.Store.<init>(Store.java:136)
	... 21 common frames omitted
ERROR 2022-09-14 18:11:34.630 [main] net.pms.PMS FATAL ERROR: Unable to bind web interface on port: 9001, because: Address already in use: bind
```
I also tried it own my system and I do get those errors too and I noticed the javaw.exe process stays open for a bit after quitting UMS.

I am sure they do not alter the .xml files and are not running in headless mode.

I think it might be better for the old logs to be renamed first, a fresh log created then zip the old ones on startup. this could help older systems (like mine) to keep up.
I don't know how this could block binding of the port though.

I also need help on the original post as I think there may be a network protocol issue I know nothing about. I saw something about "Expected soap action in request" which may be relevant.

--- a/src/main/java/net/pms/PMS.java
+++ b/src/main/java/net/pms/PMS.java
@@ -650,63 +650,7 @@ public class PMS {
 		Runtime.getRuntime().addShutdownHook(new Thread("UMS Shutdown") {
 			@Override
 			public void run() {
-				try {
-					//Stop network scanner
-					NetworkConfiguration.stop();
-
-					LOGGER.debug("Shutting down the media server");
-					MediaServer.stop();
-					Thread.sleep(500);
-
-					LOGGER.debug("Shutting down all active processes");
-
-					if (Services.processManager() != null) {
-						Services.processManager().stop();
-					}
-					for (Process p : currentProcesses) {
-						try {
-							p.exitValue();
-						} catch (IllegalThreadStateException ise) {
-							LOGGER.trace("Forcing shutdown of process: " + p);
-							ProcessUtil.destroy(p);
-						}
-					}
-				} catch (InterruptedException e) {
-					LOGGER.debug("Interrupted while shutting down..");
-					LOGGER.trace("", e);
-				}
-
-				// Destroy services
-				Services.destroy();
-
-				LOGGER.info("Stopping {} {}", PropertiesUtil.getProjectProperties().get("project.name"), getVersion());
-				/**
-				 * Stopping logging gracefully (flushing logs)
-				 * No logging is available after this point
-				 */
-				ILoggerFactory iLoggerContext = LoggerFactory.getILoggerFactory();
-				if (iLoggerContext instanceof LoggerContext) {
-					((LoggerContext) iLoggerContext).stop();
-				} else {
-					LOGGER.error("Unable to shut down logging gracefully");
-					System.err.println("Unable to shut down logging gracefully");
-				}
-
-				// Shut down library scanner
-				if (getConfiguration().getUseCache()) {
-					if (LibraryScanner.isScanLibraryRunning()) {
-						LOGGER.debug("LibraryScanner is still running, attempting to stop it");
-						LibraryScanner.stopScanLibrary();
-					} else {
-						LOGGER.debug("LibraryScanner already stopped");
-					}
-				}
-
-				if (MediaDatabase.isInstantiated()) {
-					LOGGER.debug("Shutting down database");
-					MediaDatabase.shutdown();
-					MediaDatabase.createDatabaseReportIfNeeded();
-				}
+				shutdown();
 			}
 		});
 
@@ -1149,6 +1093,79 @@ public class PMS {
 	 */
 	public static String getVersion() {
 		return PropertiesUtil.getProjectProperties().get("project.version");
+	}
+
+	/**
+	 * Shutdown the currently running Universal Media Server.
+	 */
+	public static void shutdown() {
+		try {
+			if (Platform.isWindows()) {
+				WindowsNamedPipe.setLoop(false);
+			}
+			//Stop network scanner
+			NetworkConfiguration.stop();
+
+			LOGGER.debug("Shutting down the media server");
+			MediaServer.stop();
+			Thread.sleep(500);
+
+			LOGGER.debug("Shutting down all active processes");
+
+			if (Services.processManager() != null) {
+				Services.processManager().stop();
+			}
+			for (Process p : get().currentProcesses) {
+				try {
+					p.exitValue();
+				} catch (IllegalThreadStateException ise) {
+					LOGGER.trace("Forcing shutdown of process: " + p);
+					ProcessUtil.destroy(p);
+				}
+			}
+		} catch (InterruptedException e) {
+			LOGGER.debug("Interrupted while shutting down..");
+			LOGGER.trace("", e);
+		}
+
+		// Destroy services
+		Services.destroy();
+
+		LOGGER.info("Stopping {} {}", PropertiesUtil.getProjectProperties().get("project.name"), getVersion());
+		/**
+		 * Stopping logging gracefully (flushing logs)
+		 * No logging is available after this point
+		 */
+		ILoggerFactory iLoggerContext = LoggerFactory.getILoggerFactory();
+		if (iLoggerContext instanceof LoggerContext) {
+			((LoggerContext) iLoggerContext).stop();
+		} else {
+			LOGGER.error("Unable to shut down logging gracefully");
+			System.err.println("Unable to shut down logging gracefully");
+		}
+
+		// Shut down library scanner
+		if (getConfiguration().getUseCache()) {
+			if (LibraryScanner.isScanLibraryRunning()) {
+				LOGGER.debug("LibraryScanner is still running, attempting to stop it");
+				LibraryScanner.stopScanLibrary();
+			} else {
+				LOGGER.debug("LibraryScanner already stopped");
+			}
+		}
+
+		if (MediaDatabase.isInstantiated()) {
+			LOGGER.debug("Shutting down database");
+			MediaDatabase.shutdown();
+			MediaDatabase.createDatabaseReportIfNeeded();
+		}
+	}
+
+	/**
+	 * Terminates the currently running Universal Media Server.
+	 */
+	public static void quit() {
+		System.exit(0);
 	}
 
 	/**

--- a/src/main/java/net/pms/newgui/LooksFrame.java
+++ b/src/main/java/net/pms/newgui/LooksFrame.java
@@ -46,7 +46,6 @@ import net.pms.configuration.PmsConfiguration;
 import net.pms.configuration.RendererConfiguration;
 import net.pms.gui.EConnectionState;
 import net.pms.io.BasicSystemUtils;
-import net.pms.io.WindowsNamedPipe;
 import net.pms.network.mediaserver.MediaServer;
 import net.pms.newgui.components.AnimatedIcon;
 import net.pms.newgui.components.AnimatedIcon.AnimatedIconStage;
@@ -645,7 +644,6 @@ public class LooksFrame extends JFrame implements IGui, Observer {
 	}
 
 	public void quit() {
-		WindowsNamedPipe.setLoop(false);
 		windowProperties.dispose();
 		try {
 			Thread.sleep(100);
@@ -653,7 +651,7 @@ public class LooksFrame extends JFrame implements IGui, Observer {
 			LOGGER.error("Interrupted during shutdown: {}", e);
 		}
 
-		System.exit(0);
+		PMS.quit();
 	}
 
 	@Override

--- a/src/main/java/net/pms/util/ProcessUtil.java
+++ b/src/main/java/net/pms/util/ProcessUtil.java
@@ -28,6 +28,7 @@ import java.lang.reflect.Field;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
@@ -278,8 +279,8 @@ public class ProcessUtil {
 
 	// Shutdown UMS and either reboot or run the given command (e.g. a script to
 	// restart UMS)
-	public static void reboot(ArrayList<String> cmd, Map<String, String> env, String startdir, String... umsOptions) {
-		final ArrayList<String> reboot;
+	public static void reboot(List<String> cmd, Map<String, String> env, String startdir, String... umsOptions) {
+		final List<String> reboot;
 		String macAppPath = null;
 		if (Platform.isMac()) {
 			String libraryPath = ManagementFactory.getRuntimeMXBean().getLibraryPath();
@@ -323,31 +324,30 @@ public class ProcessUtil {
 			startdir = System.getProperty("user.dir");
 		}
 
-		System.out.println("Starting: " + StringUtils.join(cmd, " "));
-
 		final ProcessBuilder pb = new ProcessBuilder(cmd);
 		if (env != null) {
 			pb.environment().putAll(env);
 		}
 		pb.directory(new File(startdir));
-		System.out.println("In folder: " + pb.directory());
 		try {
+			LOGGER.info("Starting: " + StringUtils.join(cmd, " "));
+			LOGGER.info("In folder: " + pb.directory());
+			PMS.shutdown();
 			pb.start();
 		} catch (Exception e) {
 			e.printStackTrace();
-			return;
 		}
-		System.exit(0);
+		PMS.quit();
 	}
 
 	// Reconstruct the command that started this jvm, including all options.
 	// See
 	// http://stackoverflow.com/questions/4159802/how-can-i-restart-a-java-application
 	// http://stackoverflow.com/questions/1518213/read-java-jvm-startup-parameters-eg-xmx
-	public static ArrayList<String> getUMSCommand() {
-		ArrayList<String> reboot = new ArrayList<>();
+	public static List<String> getUMSCommand() {
+		List<String> reboot = new ArrayList<>();
 		File jvmPath = new File(System.getProperty("java.home"));
-		String jvmExecutableName = Platform.isWindows() && System.console() == null ? "javaw" : "java";
+		String jvmExecutableName = Platform.isWindows() && System.console() == null ? "javaw.exe" : "java";
 		File jvmExecutable = new File(jvmPath, jvmExecutableName);
 		if (!jvmExecutable.exists() || jvmExecutable.isDirectory()) {
 			jvmPath = new File(jvmPath, "bin");

